libs/cli/src/main/java/org/elasticsearch/cli/internal/JsonPrintWriter.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cli.internal;
+
+import co.elastic.logging.EcsJsonSerializer;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.time.Clock;
+
+/**
+ * {@link PrintWriter} formatting lines as JSON using a subset of fields written by {@link EcsJsonSerializer}.
+ * Anything that looks like a JSON object is immediately passed through and written as is.
+ */
+public class JsonPrintWriter extends PrintWriter {
+    private static final int INITIAL_BUFFER_SIZE = 256;
+    private static final int MAX_BUFFER_SIZE = 4 * INITIAL_BUFFER_SIZE;
+
+    private final StringBuilder buffer = new StringBuilder(INITIAL_BUFFER_SIZE);
+    private final Clock clock;
+
+    public JsonPrintWriter(OutputStream out, boolean autoFlush) {
+        this(out, autoFlush, Clock.systemUTC());
+    }
+
+    protected JsonPrintWriter(OutputStream out, boolean autoFlush, Clock clock) {
+        super(out, autoFlush);
+        this.clock = clock;
+    }
+
+    @Override
+    public void write(String msg) {
+        if (isJsonObject(msg)) {
+            super.write(msg);
+        } else {
+            synchronized (buffer) {
+                buffer.append(msg);
+            }
+        }
+    }
+
+    @Override
+    public void println(String msg) {
+        final String json;
+        if (isJsonObject(msg)) {
+            // Forward as is, ignore current buffer.
+            json = msg;
+        } else {
+            // Don't mind if another thread might miss a none-empty buffer.
+            // It's highly unlikely this is meant to be concatenated into the same line.
+            json = toJson(buffer.isEmpty() ? msg : getAndClearBuffer(msg), clock.millis());
+        }
+        super.write(json);
+        super.println();
+    }
+
+    @Override
+    public void println() {
+        flush();
+        super.println();
+    }
+
+    @Override
+    public void flush() {
+        synchronized (buffer) {
+            if (buffer.isEmpty() == false) {
+                super.write(toJson(getAndClearBuffer(), clock.millis()));
+            }
+        }
+        super.flush();
+    }
+
+    @Override
+    public void close() {
+        flush();
+        super.close();
+    }
+
+    /**
+     * Converts {@code msg} to JSON using the {@link EcsJsonSerializer}.
+     */
+    protected static String toJson(String msg, long timeMillis) {
+        StringBuilder builder = new StringBuilder(INITIAL_BUFFER_SIZE);
```

Comment:
TBD, should this resolve / include any additional fields supported by EcsJsonSerializer?